docs/proposals/20240517-separate-yurtmanager-clients.md

```diff
@@ -0,0 +1,274 @@
+
+## Motivation
+
+1. Sharing RBAC among controller/webhooks violates the principle of least authority.
+2. Sharing user agent is hard to trace the request from audit logs.
+
+### Goals
+
+1. Separate RBAC for different controllers and webhooks and make each has its own identity in Kubernetes.
+2. Separate User agents for different controllers and webhooks.
+3. Restrict each controller/webhook to only the permissions it may use.
+
+### Non-Goals
+
+1. Divide Yurt-manager into multiple components.
+
+## Proposal
+
+### Design principals
+
+1. Compatible with currently used controller-runtime framework. Developers of yurt-manager will not notice the changes underhood.
+2. Follow the principle of least authority. One controller/webhook should not be granted permissions it will not use.
+
+### Critical Questions
+
+1. What's the granularity of division?
+
+The structure of Yurt-manager is relatively complex. Yurt-manager is made up of controllers and webhooks. Some controllers may also have several sub-controllers. Also, for RBAC division, yurt-manager itself needs some base permissions to manage the whole component.
+
+```
+yurt-manager (base)
+- controllers
+  - yurt-app-set
+  - yurt-coordiator
+    - yurt-coordinator-cert
+  - ...
+- webhooks
+  - yurt-app-set
+  - ...
+```
+
+In order to be compatible with the existing system, and also easy to manage permissions, we decide to divide the RBAC in this pattern
+
+```
+base: used in the entire life-cycle of yurt-manager, such as controller initialization and webhook server set up
+yurt-app-set(controller): used for yurt-app-set controller only
+yurt-coordinator(controller): used for yurt-coordinator-cert/pod-binding and all sub-controllers of yurt-coordinator
+yurt-app-set(webhook): used for yurt-app-set webhook only
+...
+```
+
+2. How to use different RBAC for different controller/webhooks?
+
+After different RBACs are prepared, we should make sure they are properly used by different controllers. We consider two feasible approaches here to achieve this goal.
+
+1) User impersonation
+
+Kubernets provides a [mechanism](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation) that one user can act as another user through impersonation headers. These let requests manually override the user info a request authenticates as. 
+
+2) Token override
+
+We can override the client config' token with the prepared ServiceAccount token which has been bound to prescribed roles or cluster roles.
+
+Considering the additional complexity of the first approach, we choose the second one to override the token directly when building the client for each controller/webhook.
+
+### Solution Introduction
+
+The whole solution consists of two steps: 
+
+1. Generate RBAC
+
+First, prepare RBAC artifacts including role/clusterrole, rolebindidng/clusterrolebinding and serviceaccount for controllers and webhooks in yurt-manager. The specific steps are as following:
```

Comment:
the name of clusterrole/clusterrolebinding/sa should begin with prefix `openyurt:controller:`, so it is convenient for user to know that this resources is used by controllers of openyurt.

Reply:
OK, I will take care of this in implementation.